extension/wallet/src/lib_wallet.rs

```diff
@@ -0,0 +1,348 @@
+    sign_loan::sign_loan,
+    withdraw_everything_to::withdraw_everything_to,
+};
+
+mod extract_loan;
+mod extract_trade;
+mod get_balances;
+mod get_transaction_history;
+mod make_create_swap_payload;
+mod make_loan_request;
+mod repay_loan;
+mod sign_and_send_swap_transaction;
+mod sign_loan;
+mod withdraw_everything_to;
+
+// TODO: make this configurable through extension option UI
+pub const DEFAULT_SAT_PER_VBYTE: u64 = 1;
+
+#[async_trait]
+pub trait Wallet {
+    async fn fetch_transaction(&self, txid: Txid) -> Result<Transaction>;
+    async fn fetch_transaction_history(&self, address: &Address) -> Result<Vec<Txid>>;
+    async fn get_fee_estimates(&self) -> Result<FeeEstimatesResponse>;
+    async fn broadcast(&self, tx: Transaction) -> Result<Txid>;
+    fn get_address(&self) -> Address;
+    fn blinding_key(&self) -> SecretKey;
+    fn secret_key(&self) -> SecretKey;
+    fn public_key(&self) -> PublicKey {
+        PublicKey::from_secret_key(SECP256K1, &self.secret_key())
+    }
+    /// A pure function to compute the balances of the wallet given a set of [`TxOut`]s.
+    fn compute_balances(&self, txouts: &[TxOut]) -> Vec<BalanceEntry> {
+        let grouped_txouts = txouts
+            .iter()
+            .filter_map(|utxo| match utxo {
+                TxOut {
+                    asset: confidential::Asset::Explicit(asset),
+                    value: confidential::Value::Explicit(value),
+                    ..
+                } => Some((*asset, *value)),
+                txout => match txout.unblind(SECP256K1, self.blinding_key()) {
+                    Ok(unblinded_txout) => Some((unblinded_txout.asset, unblinded_txout.value)),
+                    Err(e) => {
+                        log::warn!("failed to unblind txout: {}", e);
+                        None
+                    }
+                },
+            })
+            .into_group_map();
+
+        grouped_txouts
+            .into_iter()
+            .map(|(asset, utxos)| {
+                let total_sum = utxos.into_iter().sum();
+                BalanceEntry {
+                    asset,
+                    value: total_sum,
+                }
+            })
+            .collect()
+    }
+    async fn fetch_utxos(&self) -> Result<Vec<Utxo>>;
+    async fn get_txouts<T: Send, FM: Fn(Utxo, TxOut) -> Result<Option<T>> + Copy + Send>(
+        &self,
+        filter_map: FM,
+    ) -> Result<Vec<T>> {
+        let utxos = self.fetch_utxos().await?;
+
+        let txouts = utxos
+            .into_iter()
+            .map(move |utxo| async move {
+                let mut tx = self.fetch_transaction(utxo.txid).await?;
+                let txout = tx.output.remove(utxo.vout as usize);
+
+                filter_map(utxo, txout)
+            })
+            .collect::<FuturesUnordered<_>>()
+            .filter_map(|r| std::future::ready(r.transpose()))
+            .try_collect::<Vec<_>>()
+            .await?;
+
+        Ok(txouts)
+    }
```

Comment:
Same here, this should be standalone but with a type parameter `W: Wallet` so you can call `fetch_utxos` and `fetch_transaction`.

Alternatively, you could make this a function on an extension trait as well, the same way `FutureExt` adds functionality to everything that implements `Future`.

Default functions in traits only make sense if you want to potentially override them. There is no reason why we would override this.